File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        // edge case
        if(root == NULL) return 0;
        int ans = 1;
        queue<TreeNode*> q;
        q.emplace(root);
        while(!q.empty()){
            int n = q.size();
            for(int i = 0; i<n; i++){
                TreeNode* curr = q.front(); q.pop();
                if(curr->left != NULL) q.emplace(curr->left);
                if(curr->right != NULL) q.emplace(curr->right);
                if(curr->left == NULL && curr->right == NULL) return ans;
            }
            ans++;
        }
        return -1;
    }
};

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
 // using array to implement queue -> time O(n) for operations in this queue
var minDepth = function(root) {
    if(root === null) return 0;
    let ans = 1;
    const queue = [root];
    while(queue.length){
        const n = queue.length;
        for(let i = 0; i<n; i++){
            let curr_node = queue.shift();
            if(curr_node.left !== null) queue.push(curr_node.left);
            if(curr_node.right !== null) queue.push(curr_node.right);
            if(curr_node.left === null && curr_node.right === null) return ans;
        }
        ans++;
    }
    return -1;
};